File: archive/python-legacy/meept/memory/personality.py
"""Self-model / personality tracking for the meept bot.

Maintains a Markdown document (``personality.md``) that captures the bot's
evolving communication style, observed expertise areas, creator preferences,
and recurring themes.  When an LLM client is available the personality
description is updated automatically by asking the model to integrate new
interaction summaries.
"""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Default personality template
# ---------------------------------------------------------------------------

_DEFAULT_PERSONALITY = """\
# Meept Personality Profile

## Communication Style
- Clear, concise, and helpful
- Adapts formality to match the conversation partner
- Prefers actionable answers over vague suggestions

## Areas of Expertise Observed
- General-purpose assistance

## Creator Preferences
- No specific preferences recorded yet

## Recurring Themes
- None observed yet

## Interaction Notes
- Profile initialised; will evolve with further interactions.
"""

# ---------------------------------------------------------------------------
# LLM prompt for personality evolution
# ---------------------------------------------------------------------------

_UPDATE_SYSTEM_PROMPT = """\
You are a self-reflective assistant maintaining a personality profile document.
Given the current profile and a summary of recent interactions, produce an
UPDATED version of the profile in Markdown.

Rules:
1. Preserve the section headings exactly (## Communication Style, ## Areas of
   Expertise Observed, ## Creator Preferences, ## Recurring Themes,
   ## Interaction Notes).
2. Only add or refine bullet points -- never remove information unless it
   directly contradicts new evidence.
3. Keep bullet points concise (one line each).
4. The document should remain under 80 lines.
5. Output ONLY the updated Markdown -- no commentary or code fences.
"""


# ---------------------------------------------------------------------------
# Public class
# ---------------------------------------------------------------------------


class PersonalityModel:
    """Manages the bot's evolving self-model as a Markdown personality file.

    Parameters
    ----------
    data_dir:
        Directory where ``personality.md`` lives.
    llm_client:
        Optional :class:`~meept.llm.client.LLMClient` used to evolve the
        profile via summarisation.  Without it, the profile is static.
    """

    def __init__(self, data_dir: Path, llm_client: Any = None) -> None:
        self._data_dir = data_dir
        self._llm_client = llm_client
        self._file_path = data_dir / "personality.md"
        self._description: str = _DEFAULT_PERSONALITY
        self._interaction_count: int = 0
        self._last_updated: datetime | None = None

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def load(self) -> None:
        """Load the personality profile from disk, or create the default."""
        self._data_dir.mkdir(parents=True, exist_ok=True)

        if self._file_path.exists():
            self._description = self._file_path.read_text(encoding="utf-8")
            logger.info("Loaded personality profile from %s.", self._file_path)
        else:
            self._description = _DEFAULT_PERSONALITY
            await self.save()
            logger.info("Created default personality profile at %s.", self._file_path)

    # ------------------------------------------------------------------
    # Update
    # ------------------------------------------------------------------

    async def update(self, interaction_summary: str) -> None:
        """Integrate a new interaction summary into the personality profile.

        If an LLM client is available the model is asked to rewrite the
        profile to incorporate the new evidence.  Otherwise the summary is
        appended verbatim under **Interaction Notes**.

        Parameters
        ----------
        interaction_summary:
            A short textual summary of a recent interaction or set of
            interactions (e.g. ``"User asked about Python asyncio patterns;
            prefers code examples over prose."``).
        """
        self._interaction_count += 1
        now_iso = datetime.now(timezone.utc).isoformat(timespec="seconds")

        if self._llm_client is not None:
            updated = await self._update_via_llm(interaction_summary)
            if updated:
                self._description = updated
                self._last_updated = datetime.now(timezone.utc)
                await self.save()
                logger.info("Personality profile updated via LLM (interaction #%d).", self._interaction_count)
                return

        # Fallback: append the summary under Interaction Notes.
        note_line = f"- [{now_iso}] {interaction_summary}"
        self._description = _append_to_section(
            self._description, "## Interaction Notes", note_line
        )
        self._last_updated = datetime.now(timezone.utc)
        await self.save()
        logger.info(
            "Personality profile updated with manual note (interaction #%d).",
            self._interaction_count,
        )

    # ------------------------------------------------------------------
    # Accessors
    # ------------------------------------------------------------------

    def get_description(self) -> str:
        """Return the current personality description as Markdown."""
        return self._description

    @property
    def interaction_count(self) -> int:
        """Number of ``update()`` calls since last load."""
        return self._interaction_count

    @property
    def last_updated(self) -> datetime | None:
        """Timestamp of the most recent ``update()`` call."""
        return self._last_updated

    # ------------------------------------------------------------------
    # Persistence
    # ------------------------------------------------------------------

    async def save(self) -> None:
        """Persist the current personality profile to disk."""
        self._data_dir.mkdir(parents=True, exist_ok=True)
        self._file_path.write_text(self._description, encoding="utf-8")
        logger.debug("Saved personality profile to %s.", self._file_path)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    async def _update_via_llm(self, interaction_summary: str) -> str | None:
        """Ask the LLM to produce an updated personality profile.

        Returns the new Markdown string on success, or ``None`` on failure.
        """
        # Import here to avoid circular dependency at module level.
        from meept.llm.models import ChatMessage, Role

        messages = [
            ChatMessage(role=Role.SYSTEM, content=_UPDATE_SYSTEM_PROMPT),
            ChatMessage(
                role=Role.USER,
                content=(
                    f"## Current Profile\n\n{self._description}\n\n"
                    f"---\n\n"
                    f"## Recent Interaction Summary\n\n{interaction_summary}\n\n"
                    f"---\n\n"
                    f"Please produce the updated personality profile."
                ),
            ),
        ]

        try:
            response = await self._llm_client.chat(messages, temperature=0.4, max_tokens=2048)
            content = response.content
            if content and content.strip():
                return content.strip()
        except Exception:
            logger.warning(
                "LLM-based personality update failed -- falling back to manual append.",
                exc_info=True,
            )
        return None


# ---------------------------------------------------------------------------
# Module-level helpers
# ---------------------------------------------------------------------------


def _append_to_section(document: str, heading: str, line: str) -> str:
    """Append *line* after the last bullet under *heading* in a Markdown doc.

    If the heading is not found, the line is appended at the end.
    """
    lines = document.split("\n")
    insert_index: int | None = None
    in_section = False

    for i, raw_line in enumerate(lines):
        stripped = raw_line.strip()
        if stripped == heading:
            in_section = True
            insert_index = i + 1
            continue
        if in_section:
            # Another heading means we left the target section.
            if stripped.startswith("## "):
                break
            # Track last non-empty line within the section.
            if stripped:
                insert_index = i + 1

    if insert_index is not None:
        lines.insert(insert_index, line)
    else:
        # Heading not found -- append at end.
        lines.append("")
        lines.append(heading)
        lines.append(line)

    return "\n".join(lines)

File: internal/memory/personality.go
package memory

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

const defaultPersonality = `# Meept Personality Profile

## Communication Style
- Clear, concise, and helpful
- Adapts formality to match the conversation partner
- Prefers actionable answers over vague suggestions

## Areas of Expertise Observed
- General-purpose assistance

## Creator Preferences
- No specific preferences recorded yet

## Recurring Themes
- None observed yet

## Interaction Notes
- Profile initialised; will evolve with further interactions.
`

// PersonalityMemory manages the bot's evolving self-model as a Markdown file.
// It tracks communication style, expertise areas, user preferences, and
// interaction patterns.
type PersonalityMemory struct {
	dataDir          string
	filePath         string
	description      string
	interactionCount int
	lastUpdated      *time.Time
	mu               sync.RWMutex
	logger           *slog.Logger
}

// PersonalityMemoryConfig holds configuration for personality memory.
type PersonalityMemoryConfig struct {
	// DataDir is the directory for the personality file.
	DataDir string
	// Logger for operations.
	Logger *slog.Logger
}

// NewPersonalityMemory creates a new personality memory instance.
func NewPersonalityMemory(cfg PersonalityMemoryConfig) *PersonalityMemory {
	if cfg.Logger == nil {
		cfg.Logger = slog.Default()
	}
	return &PersonalityMemory{
		dataDir:     cfg.DataDir,
		filePath:    filepath.Join(cfg.DataDir, "personality.md"),
		description: defaultPersonality,
		logger:      cfg.Logger,
	}
}

// Load loads the personality profile from disk, or creates the default.
func (p *PersonalityMemory) Load(ctx context.Context) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	// Ensure directory exists
	if err := os.MkdirAll(p.dataDir, 0755); err != nil {
		return fmt.Errorf("failed to create data directory: %w", err)
	}

	// Try to load existing file
	data, err := os.ReadFile(p.filePath)
	if err != nil {
		if os.IsNotExist(err) {
			// Create default profile
			p.description = defaultPersonality
			if err := p.saveUnlocked(); err != nil {
				return err
			}
			p.logger.Info("Created default personality profile", "path", p.filePath)
			return nil
		}
		return fmt.Errorf("failed to read personality file: %w", err)
	}

	p.description = string(data)
	p.logger.Info("Loaded personality profile", "path", p.filePath)
	return nil
}

// Update integrates new interaction data into the personality profile.
// The summary should describe recent interaction patterns or preferences.
func (p *PersonalityMemory) Update(ctx context.Context, interactionSummary string) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	p.interactionCount++
	nowISO := time.Now().UTC().Format(time.RFC3339)

	// Append the summary under Interaction Notes
	noteLine := fmt.Sprintf("- [%s] %s", nowISO, interactionSummary)
	p.description = appendToSection(p.description, "## Interaction Notes", noteLine)

	now := time.Now()
	p.lastUpdated = &now

	if err := p.saveUnlocked(); err != nil {
		return err
	}

	p.logger.Info("Personality profile updated",
		"interaction_count", p.interactionCount,
	)
	return nil
}

// UpdateKey updates or adds a specific key-value pair in the personality.
// This is useful for tracking specific preferences.
func (p *PersonalityMemory) UpdateKey(ctx context.Context, section string, key string, value string) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	// Find or create the section and update/add the key
	entry := fmt.Sprintf("- %s: %s", key, value)
	p.description = updateInSection(p.description, "## "+section, key, entry)

	now := time.Now()
	p.lastUpdated = &now

	return p.saveUnlocked()
}

// GetDescription returns the current personality description as Markdown.
func (p *PersonalityMemory) GetDescription() string {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.description
}

// GetPersonality returns the personality as a parsed map of sections.
func (p *PersonalityMemory) GetPersonality() map[string][]string {
	p.mu.RLock()
	defer p.mu.RUnlock()

	return parsePersonality(p.description)
}

// InteractionCount returns the number of updates since load.
func (p *PersonalityMemory) InteractionCount() int {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.interactionCount
}

// LastUpdated returns the timestamp of the most recent update.
func (p *PersonalityMemory) LastUpdated() *time.Time {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.lastUpdated
}

// Save persists the current personality profile to disk.
func (p *PersonalityMemory) Save() error {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.saveUnlocked()
}

func (p *PersonalityMemory) saveUnlocked() error {
	if err := os.MkdirAll(p.dataDir, 0755); err != nil {
		return fmt.Errorf("failed to create data directory: %w", err)
	}
	if err := os.WriteFile(p.filePath, []byte(p.description), 0644); err != nil {
		return fmt.Errorf("failed to write personality file: %w", err)
	}
	p.logger.Debug("Saved personality profile", "path", p.filePath)
	return nil
}

// Close saves the personality profile.
func (p *PersonalityMemory) Close() error {
	return p.Save()
}

// appendToSection appends a line after the last bullet under a heading.
func appendToSection(document, heading, line string) string {
	lines := strings.Split(document, "\n")
	var insertIndex int
	inSection := false

	for i, rawLine := range lines {
		stripped := strings.TrimSpace(rawLine)
		if stripped == heading {
			inSection = true
			insertIndex = i + 1
			continue
		}
		if inSection {
			// Another heading means we left the target section
			if strings.HasPrefix(stripped, "## ") {
				break
			}
			// Track last non-empty line within the section
			if stripped != "" {
				insertIndex = i + 1
			}
		}
	}

	if insertIndex > 0 {
		// Insert the line
		newLines := make([]string, 0, len(lines)+1)
		newLines = append(newLines, lines[:insertIndex]...)
		newLines = append(newLines, line)
		newLines = append(newLines, lines[insertIndex:]...)
		return strings.Join(newLines, "\n")
	}

	// Heading not found - append at end
	return document + "\n\n" + heading + "\n" + line
}

// updateInSection updates or adds an entry in a section.
func updateInSection(document, heading, key, newEntry string) string {
	lines := strings.Split(document, "\n")
	var _, sectionEnd int
	inSection := false
	keyFound := false

	for i, rawLine := range lines {
		stripped := strings.TrimSpace(rawLine)
		if stripped == heading {
			inSection = true
			_ = i + 1 // sectionStart
			continue
		}
		if inSection {
			if strings.HasPrefix(stripped, "## ") {
				sectionEnd = i
				break
			}
			// Check if this line contains the key
			if strings.Contains(stripped, key+":") || strings.HasPrefix(stripped, "- "+key) {
				lines[i] = newEntry
				keyFound = true
			}
			sectionEnd = i + 1
		}
	}

	if keyFound {
		return strings.Join(lines, "\n")
	}

	// Key not found - add it to the section
	if sectionEnd > 0 {
		newLines := make([]string, 0, len(lines)+1)
		newLines = append(newLines, lines[:sectionEnd]...)
		newLines = append(newLines, newEntry)
		newLines = append(newLines, lines[sectionEnd:]...)
		return strings.Join(newLines, "\n")
	}

	// Section not found - create it
	return document + "\n\n" + heading + "\n" + newEntry
}

// parsePersonality parses a personality document into sections.
func parsePersonality(document string) map[string][]string {
	result := make(map[string][]string)
	lines := strings.Split(document, "\n")
	currentSection := ""

	for _, line := range lines {
		stripped := strings.TrimSpace(line)
		if after, ok := strings.CutPrefix(stripped, "## "); ok {
			currentSection = after
			result[currentSection] = []string{}
			continue
		}
		if currentSection != "" && strings.HasPrefix(stripped, "- ") {
			entry := strings.TrimPrefix(stripped, "- ")
			result[currentSection] = append(result[currentSection], entry)
		}
	}

	return result
}

// PersonalityStore provides a key-value interface over personality data.
type PersonalityStore struct {
	memory *PersonalityMemory
}

// NewPersonalityStore wraps a PersonalityMemory with key-value access.
func NewPersonalityStore(memory *PersonalityMemory) *PersonalityStore {
	return &PersonalityStore{memory: memory}
}

// Get retrieves a preference value.
func (s *PersonalityStore) Get(ctx context.Context, key string) (string, error) {
	if s.memory == nil {
		return "", errors.New("personality memory not initialized")
	}

	personality := s.memory.GetPersonality()
	for _, entries := range personality {
		for _, entry := range entries {
			if after, ok := strings.CutPrefix(entry, key+":"); ok {
				return strings.TrimSpace(after), nil
			}
		}
	}

	return "", nil // Not found, not an error
}

// Set stores a preference value.
func (s *PersonalityStore) Set(ctx context.Context, key, value string) error {
	if s.memory == nil {
		return errors.New("personality memory not initialized")
	}

	return s.memory.UpdateKey(ctx, "Creator Preferences", key, value)
}

// GetAll returns all personality data as a map.
func (s *PersonalityStore) GetAll(ctx context.Context) (map[string][]string, error) {
	if s.memory == nil {
		return nil, errors.New("personality memory not initialized")
	}

	return s.memory.GetPersonality(), nil
}
